vault/expiration.go

```diff
@@ -1213,79 +1307,133 @@ func (m *ExpirationManager) deleteEntry(leaseID string) error {
 	return nil
 }
 
-// createIndexByToken creates a secondary index from the token to a lease entry
-func (m *ExpirationManager) createIndexByToken(token, leaseID string) error {
-	saltedID, err := m.tokenStore.SaltID(m.quitContext, token)
+// hmacIndexPath takes in a token ID and lease ID and returns the storage path
+// to be used for the lease's secondary index. The storage path is created by
+// SHA2-256 HMACing the IDs using the token store's salt.
+func (m *ExpirationManager) hmacIndexPath(tokenID, leaseID string) (string, error) {
+	tokenIDHMAC, err := m.tokenStore.hmac(m.quitContext, tokenID)
 	if err != nil {
-		return err
+		return "", err
 	}
 
+	leaseIDHMAC, err := m.tokenStore.hmac(m.quitContext, leaseID)
+	if err != nil {
+		return "", err
+	}
+
+	return path.Join(tokenIDHMAC, leaseIDHMAC), nil
+}
+
+// Deprecated: hashIndexPath takes in a token ID and lease ID and returns the
+// storage path to be used for the lease's secondary index. The storage path is
+// created by SHA1 hashing the IDs using the token store's salt. This is only
+// here for backwards compatibility.
+func (m *ExpirationManager) hashIndexPath(tokenID, leaseID string) (string, error) {
+	saltedID, err := m.tokenStore.SaltID(m.quitContext, tokenID)
+	if err != nil {
+		return "", err
+	}
 	leaseSaltedID, err := m.tokenStore.SaltID(m.quitContext, leaseID)
+	if err != nil {
+		return "", err
+	}
+
+	return path.Join(saltedID, leaseSaltedID), nil
+}
+
+// createIndexByToken creates a secondary index from the token to a lease entry
+func (m *ExpirationManager) createIndexByToken(tokenID, leaseID string) error {
+	indexPath, err := m.hmacIndexPath(tokenID, leaseID)
 	if err != nil {
 		return err
 	}
 
-	ent := logical.StorageEntry{
-		Key:   saltedID + "/" + leaseSaltedID,
+	entry := logical.StorageEntry{
+		Key:   indexPath,
 		Value: []byte(leaseID),
 	}
-	if err := m.tokenView.Put(m.quitContext, &ent); err != nil {
+
+	err = m.tokenView.Put(m.quitContext, &entry)
+	if err != nil {
 		return errwrap.Wrapf("failed to persist lease index entry: {{err}}", err)
 	}
+
 	return nil
 }
 
 // indexByToken looks up the secondary index from the token to a lease entry
-func (m *ExpirationManager) indexByToken(token, leaseID string) (*logical.StorageEntry, error) {
-	saltedID, err := m.tokenStore.SaltID(m.quitContext, token)
+func (m *ExpirationManager) indexByToken(tokenID, leaseID string) (*logical.StorageEntry, error) {
+	indexPath, err := m.hmacIndexPath(tokenID, leaseID)
 	if err != nil {
 		return nil, err
 	}
 
-	leaseSaltedID, err := m.tokenStore.SaltID(m.quitContext, leaseID)
+	entry, err := m.tokenView.Get(m.quitContext, indexPath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to look up secondary index entry")
+	}
+	if entry != nil {
+		return entry, nil
+	}
+
+	// This is only here for backwards compatibility
```

Comment:
Here, and in all places where you do backwards compatibility checks, is it possible to gate whether the check occurrs on whether or not any pre-verison-2 leases were found?

Obviously this relies on two things: turning the atomic value into a tri state (so, found, unsure because still running and no persisted value in which case do backwards compat, not found); and, that anywhere that you're doing this backwards compat can actually properly be gated on that value. But my thinking is it can and could save some storage calls/request time.

Reply:
I think this is a good suggestion. Although, I think that we don't necessarily need a tri-state. Both in case of `found` and `unsure` we need to do a backwards compat, and in case of `not-found` we skip; we could very well do this with two states.

Reply:
Fixed.